### PR TITLE
Re-Implementation of Gaussian noise

### PR DESCRIPTION
**Pull request by msz-rai**

### CODE REVIEW

File: src/gpu/gaussianNoiseKernels.cu [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+	float distanceInducedStDev = distance * stDevRisePerMeter;
+	float totalStDev = distanceInducedStDev + stDevBase;
+
+	float distanceError = mean + curand_normal(&randomStates[tid]) * totalStDev;
+
+	if (outDistances != nullptr) {
+		outDistances[tid] = distance + distanceError;
+	}
+
+	outPoints[tid] = toOriginTransform.inverse() * (originPoint.normalize() * (distance + distanceError));
```
Comment (PiotrMrozikRobotec) [inline src/gpu/gaussianNoiseKernels.cu:72]: I believe that can be done faster like:
outPoint = inPoint +- toOriginTransform * unitVector * distanceError

File: test/src/gaussianStressTest.cpp [hunk 2]
```diff
@@ -0,0 +1,146 @@
+#include <gtest/gtest.h>
```
Comment (PiotrMrozikRobotec) [inline test/src/gaussianStressTest.cpp:1]: Ok, why U called them Stress test?

File: include/rgl/api/core.h [hunk 3]
```diff
@@ -451,6 +461,45 @@ rgl_node_points_downsample(rgl_node_t* node, float leaf_size_x, float leaf_size_
...
+ * Creates or modifies GaussianNoiseAngularRayNode.
+ * Applies angular noise to the rays before raycasting.
+ * Graph input: rays
+ * Graph output: rays
+ * @param node If (*node) == nullptr, a new node will be created. Otherwise, (*node) will be modified.
+ * @param mean Angular noise mean in radians.
+ * @param st_dev Angular noise standard deviation in radians.
+ * @param axis Axis on which angular noise will be perform.
+ */
+RGL_API rgl_status_t
```
[the remark this answers is not in the document]
Comment (msz-rai) [reply include/rgl/api/core.h:474]: Do you mean [markdown file](https://github.com/RobotecAI/RobotecGPULidar/blob/feature/gaussian-noise-nodes/docs/GaussianNoise.md)?

File: src/gpu/kernelUtils.hpp [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <macros/cuda.hpp>
+
+#define LIMIT(count) const int tid = (blockIdx.x * blockDim.x + threadIdx.x); do {if (tid >= count) { return; }} while(false)
```
[the remark this answers is not in the document]
Comment (msz-rai) [reply src/gpu/kernelUtils.hpp:17]: I just moved this macro to a separate file :wink:

[the remark this answers is not in the document]
Comment (PiotrMrozikRobotec) [reply src/gpu/kernelUtils.hpp:17]: I know :P

File: include/rgl/api/core.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (PiotrMrozikRobotec) [reply include/rgl/api/core.h:474]: Yep

File: src/math/Mat3x4f.hpp [hunk 5]
```diff
@@ -79,13 +68,21 @@ struct Mat3x4f
 		return rz * ry * rx;
 	}
 
-	static inline Mat3x4f rotation(float x, float y, float z)
+	static HostDevFn inline Mat3x4f rotationRad(rgl_axis_t axis, float angleRad)
+	{
+		float rotX = axis == RGL_AXIS_X ? angleRad : 0.0f;
+		float rotY = axis == RGL_AXIS_Y ? angleRad : 0.0f;
+		float rotZ = axis == RGL_AXIS_Z ? angleRad : 0.0f;
+		return rotationRad(rotX, rotY, rotZ);
```
[the remark this answers is not in the document]
Comment (msz-rai) [reply src/math/Mat3x4f.hpp:76]: Could you provide the source of this code?

[the remark this answers is not in the document]
Comment (PiotrMrozikRobotec) [reply src/math/Mat3x4f.hpp:76]: https://github.com/microsoft/DirectXMath/blob/main/Inc/DirectXMathMatrix.inl

Line 1697, but You can read whole scope for rotations

File: src/gpu/gaussianNoiseKernels.cu (hunk 1 above)
Comment (msz-rai) [reply src/gpu/gaussianNoiseKernels.cu:72]: Your solution doesn't pass tests. I have found optimization like that:
outPoints[tid] = inPoints[tid] + inPoints[tid].normalize() * distanceError;

File: test/src/gaussianStressTest.cpp (hunk 2 above)
Comment (msz-rai) [reply test/src/gaussianStressTest.cpp:1]: This test runs gaussian noise over time (more than one iteration) and checks whether the values are the same. In the previous implementation, we had a bug in that the point cloud expanded (gaussian noise had been applied on the last noised point cloud).

Comment (PiotrMrozikRobotec) [reply test/src/gaussianStressTest.cpp:1]: Ok, so this is a large portion of repetition of the unit test, not the stress test.

Comment (msz-rai) [reply test/src/gaussianStressTest.cpp:1]: Resolved offline.

**Pull request merged**